### PR TITLE
Lint casting signed integers smaller than `isize` to raw pointers

### PR DESCRIPTION
**Pull request by oli-obk**
cc #43291 (keeping open as tracking issue)

### CODE REVIEW

**Comment by eddyb** [comment]
Whoa whoa what is all this new code? I expected a trivial check to be used only for a cargobomb run.

**Comment by eddyb** [comment]
cc @rust-lang/compiler

**Comment by oli-obk** [comment]
This code exists so we don't produce a suggestion like `5 + 6 as usize as *const T` if we had `(5 + 6) as *const T` when starting out. It is 1:1 copied from clippy.

**Comment by eddyb** [comment]
> It is 1:1 copied from clippy.

Ah, great! Would've been less surprising to see that in the PR description. However, I'd still start small, with an error-by-default lint with no interesting suggestions, for cargobomb.

**Comment by oli-obk** [comment]
done

File: src/librustc_lint/builtin.rs [hunk 1]
```diff
@@ -1265,3 +1265,57 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnionsWithDropFields {
...
+    fn check_expr(&mut self, cx: &LateContext, expr: &hir::Expr) {
+        if let hir::ExprCast(ref base, _) = expr.node {
+            let base_ty = cx.tables.expr_ty(base);
+            let target_ty = cx.tables.expr_ty(expr);
+            if let ty::TyRawPtr(..) = target_ty.sty {
+                match base_ty.sty {
+                    ty::TyInt(ast::IntTy::Is) => return,
+                    ty::TyInt(ast::IntTy::I128) |
+                    ty::TyInt(ast::IntTy::I64) => {
+                        if cx.tcx.data_layout.pointer_size.bytes() >= 8 {
```
Comment (ollie27) [inline src/librustc_lint/builtin.rs]: This doesn't seem right. I think this should be `<=` and the `if` needs to be part of the match pattern to allow this to fall through:

```rust
ty::TyInt(ast::IntTy::I128) |
ty::TyInt(ast::IntTy::I64) if cx.tcx.data_layout.pointer_size.bytes() <= 8 => {
    return;
},
ty::TyInt(ast::IntTy::I32) if cx.tcx.data_layout.pointer_size.bytes() <= 4 => {
    return;
},
ty::TyInt(ast::IntTy::I16) if cx.tcx.data_layout.pointer_size.bytes() <= 2 => {
    return;
},
```

File: src/test/run-pass/mir_misc_casts.rs [hunk 2]
```diff
@@ -120,7 +120,7 @@ fn from_1i8()
     let c10 = 1_i8 as u64;
     let c11 = 1_i8 as f32;
     let c12 = 1_i8 as f64;
-    let c13 = 1_i8 as *const libc::FILE;
+    let c13 = 1_i8 as isize as *const libc::FILE;
```
Comment (eddyb) [inline src/test/run-pass/mir_misc_casts.rs]: The correct cast btw is through `usize`. Although... I'd leave the tests unchanged and just allow the lint in them.

**Review by ollie27 — COMMENTED** [review]
There is some overlap between this and #43339 btw.

File: src/librustc_lint/builtin.rs (hunk 1 above)
Comment (eddyb) [inline src/librustc_lint/builtin.rs]: Why isn't this comparing the `ity` in `ty::TyInt(ity)` to `tcx.sess.target.int_type`?

Comment (oli-obk) [inline src/librustc_lint/builtin.rs]: thanks! you're right of course, I got the logic reversed.

File: src/librustc_lint/builtin.rs [hunk 3]
```diff
@@ -1265,3 +1265,54 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnionsWithDropFields {
...
+                    },
+                    ty::TyInt(ast::IntTy::I16) => {
+                        if cx.tcx.data_layout.pointer_size.bytes() <= 2 {
+                            return;
+                        }
+                    },
+                    // these we report
+                    ty::TyInt(_) => {},
+                    // only int casts are relevant
+                    _ => {},
```
Comment (ollie27) [inline src/librustc_lint/builtin.rs]: I assume there's supposed to be a `return` here.

**Comment by shepmaster** [comment]
@eddyb I assume you will be starting the cargobomb?

**Comment by frewsxcv** [comment]
Looks like there's an error in the book. I opened a PR upstream: https://github.com/rust-lang/book/pull/866.

**Comment by frewsxcv** [comment]
Upstream PR has merged. Updating the in-tree book submodule now: https://github.com/rust-lang/rust/pull/43688.

**Comment by oli-obk** [comment]
Thanks for taking care of that!

**Comment by frewsxcv** [comment]
@oli-obk the book should be good now, you should rebase or merge in master

**Comment by eddyb** [comment]
@frewsxcv Always rebase 😛!

**Comment by oli-obk** [comment]
ping @eddyb

**Comment by eddyb** [comment]
cc @rust-lang/infra Cargobomb crater requested.

**Comment by aidanhs** [comment]
I'll get this cargobomb run going.

**Comment by aidanhs** [comment]
Cargobomb run failed, the disk filled up because I failed to do due diligence on disk space beforehand - sorry! I'll prepare a new run now.

**Comment by aidanhs** [comment]
Cargobomb complete: http://cargobomb-reports.s3.amazonaws.com/pr-43641/index.html

**Comment by eddyb** [comment]
@oli-obk So the thing we were worried about was whether any of these would *need* the (buggy) zero-extension behavior, to know if we can just fix it right away.
Specifically, if there are negative integers, less wide than pointers, being cast to pointers.

**Comment by oli-obk** [comment]
I have checked everything but `rsjs` for which I was not able to find a repository. None of them actually depend on the buggy casting, since they only have positive values or `0`.

**Comment by oli-obk** [comment]
Closing in favour of simply making the backwards incompatible fix

**Pull request closed without merging**